libvast/vast/format/syslog.hpp

```diff
@@ -13,17 +13,265 @@
+//      BOM             = %xEF.BB.BF
+//      UTF-8-STRING    = *OCTET ; UTF-8 string as specified
+//                        ; in [RFC 3629](https://tools.ietf.org/html/rfc3629)
+//
+//      OCTET           = %d00-255
+//      SP              = %d32
+//      PRINTUSASCII    = %d33-126
+//      NONZERO-DIGIT   = %d49-57
+//      DIGIT           = %d48 / NONZERO-DIGIT
+//      NILVALUE        = "-"
+
+// Example messages: 
+// <34>1 2003-10-11T22:14:15.003Z mymachine.example.com su - ID47 - BOM'su root' failed for lonvick on /dev/pts/8
+// <165>1 2003-10-11T22:14:15.003Z mymachine.example.com evntslog - ID47 [exampleSDID@32473 iut="3" eventSource="Application" eventID="1011"] BOMAn application event log entry... 
+// <165>1 2003-10-11T22:14:15.003Z mymachine.example.com evntslog - ID47 [exampleSDID@32473 iut="3" eventSource="Application" eventID="1011"][examplePriority@32473 class="high"]      
+
+
+
+// I defined these ranges to make the parser better readable
+namespace ranges {
+  size_t prival_length_min = 0;
+  size_t prival_length_max = 3;
+  size_t prival_value_max = 191;
+
+  size_t version_length_min = 0;
+  size_t version_length_max = 2;
+  size_t version_non_zerp = 0;
+
+  size_t name_length_min = 0; // describes the minimum length for HOSTNAME, APP-NAME, PROCID, MSGID
+  size_t hostnamename_length_max = 255;
+  size_t app_name_length_max = 48;
+  size_t procid_length_max = 128;
+  size_t msgid_length_max = 32;
+
+  size_t sdname_length_min = 1;
+  size_t sd_name_length_max = 32;
+
+  size_t octet_range_min = -128;
+  size_t octet_range_max = 127;
+
+  size_t printusascii_range_min = 33;
+  size_t printusascii_range_max = 126;
+
+} // ranges
+
+struct syslog_msg_content{
+  std::string msg_content;
+};
+
+struct sd_element_param{
+  std::string name;
+  std::string value;
+};
+
+struct sd_element{
+  std::string id;
+  std::vector<sd_element_param> param;
+};
+
+struct syslog_structured_data{
+  caf::optional<std::vector<sd_element>> elements;
+};
+
+struct syslog_header{
+  uint16_t priority;
+  uint16_t version;
+  caf::optional<time> ts; // vast time
+  caf::optional<std::string> hostname;
+  caf::optional<std::string> app_name;
+  caf::optional<std::string> process_id;
+  caf::optional<std::string> msg_id;
+};
+
+struct syslog_msg{
+  syslog_header header;
+  syslog_structured_data structured_data;
+  caf::optional<syslog_msg_content> msg;
+};
+
+// Helper functions to parse attributes.
+
+// helper that parses the MSG part of a syslog message
+struct msg_content_parser : parser<msg_content_parser>{
+  using attribute = syslog_msg_content;
+
+  template <class Iterator>
+  bool parse(Iterator& f, Iterator& l, attribute& a) const{
+    using namespace parsers;
+    using namespace parser_literals;
+
+    auto octet = char_range_parser<ranges::octet_range_min, ranges::octet_range_min>{}; // 7 Bit USASCII
+     
+    auto msg_any = *(octet);
+
+    auto bom = "ï»¿"_p;
+    auto msg_utf_8 = ignore(bom) >> *(octet);
```

Comment:
```suggestion
    auto msg_utf_8 = bom >> *(octet);
```

The `ignore` shouldn't be necessary here. Literals are always ignored, see `operator""_p` in `libvast/vast/concept/parseable/core/literal.hpp`.